xdem/spatialstats.py

```diff
@@ -820,213 +1028,384 @@ def vgm_sum(h, *args):
+
+    # Define the covariance sum function times the spatial lag, for later integration
     def hcov_sum(h):
         fn = 0
-        for vario in list_vgm:
-            crange, model, psill = vario
-            fn += h*(cov(h, crange, model=model, psill=psill))
+        for i in np.arange((len(params_vgm))):
+            model_name = _get_scikitgstat_vgm_model_name(params_vgm['model'].values[i])
+            range = params_vgm['range'].values[i]
+            psill = params_vgm['psill'].values[i]
+            model_function = getattr(skg.models, model_name)
+            # For models that expect 2 parameters
+            if model_name in ['spherical', 'gaussian', 'exponential', 'cubic']:
+                # The covariance is the partial sill minus the variogram
+                fn += h * (psill - model_function(h, range, psill))
+                i += 2
+            # For models that expect 3 parameters
+            elif model_name in ['stable', 'matern']:
+                smooth = params_vgm['smooth'].values[i]
+                # The covariance is the partial sill minus the variogram
+                fn += h * (psill - model_function(h, range, psill, smooth))
+                i += 3
 
         return fn
 
+    # Get a radius for which the circle as the defined area
     h_equiv = np.sqrt(area / np.pi)
 
-    full_int = integrate_fun(hcov_sum, 0, h_equiv)
-    std_err = np.sqrt(2*np.pi*full_int / area)
+    # Integrate the covariance function between the center and the radius
+    full_int = _integrate_fun(hcov_sum, 0, h_equiv)
+
+    # Get the standard error, and return the number of effective samples
+    squared_se = 2*np.pi*full_int / area
 
-    return psill_tot/std_err**2
+    # The number of effective sample is the fraction of total sill by squared standard error
+    neff = total_sill/squared_se
 
+    return neff
 
-def neff_rect(area: float, width: float, crange1: float, psill1: float, model1: str = 'Sph', crange2: float = None,
-              psill2: float = None, model2: str = None) -> float:
+
+def neff_exact(coords: np.ndarray, errors: np.ndarray, params_vgm: pd.DataFrame, vectorized=True) -> float:
     """
-    Number of effective samples derived from numerical integration for a sum of 2 variogram functions over a rectangular area
-
-    :param area: area
-    :param width: width of rectangular area
-    :param crange1: correlation range of first variogram
-    :param psill1: partial sill of first variogram
-    :param model1: model of first variogram
-    :param crange2: correlation range of second variogram
-    :param psill2: partial sill of second variogram
-    :param model2: model of second variogram
-
-    :returns: number of effective samples
+     Exact number of effective samples derived from a double sum of covariance with euclidean coordinates based on
+     the provided variogram parameters. This method works for any shape of area.
+
+    :param coords: Coordinates at the scale of the spatial support (typically, pixel)
```

Comment:
I don't fully understand what these two arrays are. Do you expect 1D, 2D, or any dimension? Is `errors` the error calculated for the standardization, or what?